crates/iceberg/src/writer/file_writer/parquet_writer.rs

```diff
@@ -78,106 +89,407 @@ impl<T: LocationGenerator, F: FileNameGenerator> FileWriterBuilder for ParquetWr
+        self.field_id = field.id;
+        Ok(())
+    }
+
+    fn after_map_value(&mut self, _field: &NestedFieldRef) -> Result<()> {
+        self.field_names.pop();
+        Ok(())
+    }
+
+    fn schema(&mut self, _schema: &Schema, _value: Self::T) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn field(&mut self, _field: &NestedFieldRef, _value: Self::T) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn r#struct(&mut self, _struct: &StructType, _results: Vec<Self::T>) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn list(&mut self, _list: &ListType, _value: Self::T) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn map(&mut self, _map: &MapType, _key_value: Self::T, _value: Self::T) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn primitive(&mut self, _p: &PrimitiveType) -> Result<Self::T> {
+        let full_name = self.field_names.iter().map(String::as_str).join(".");
+        let field_id = self.field_id;
+        if let Some(existing_field_id) = self.name_to_id.get(full_name.as_str()) {
+            return Err(Error::new(ErrorKind::DataInvalid, format!("Invalid schema: multiple fields for name {full_name}: {field_id} and {existing_field_id}")));
+        } else {
+            self.name_to_id.insert(full_name, field_id);
+        }
+
+        Ok(())
+    }
+}
+
 /// `ParquetWriter`` is used to write arrow data into parquet file on storage.
 pub struct ParquetWriter {
+    schema: SchemaRef,
     out_file: OutputFile,
     writer: AsyncArrowWriter<AsyncFileWriter<TrackWriter>>,
     written_size: Arc<AtomicI64>,
     current_row_num: usize,
-    field_ids: Vec<i32>,
+}
+
+/// Used to aggregate min and max value of each column.
+struct MinMaxColAggregator {
+    lower_bounds: HashMap<i32, Datum>,
+    upper_bounds: HashMap<i32, Datum>,
+    schema: SchemaRef,
+}
+
+impl MinMaxColAggregator {
+    fn new(schema: SchemaRef) -> Self {
+        Self {
+            lower_bounds: HashMap::new(),
+            upper_bounds: HashMap::new(),
+            schema,
+        }
+    }
+
+    fn update(&mut self, col_id: i32, value: Statistics) -> Result<()> {
+        let crate::spec::Type::Primitive(ty) = &self
+            .schema
+            .field_by_id(col_id)
+            .ok_or_else(|| {
+                Error::new(
+                    crate::ErrorKind::Unexpected,
+                    "Failed to get field by id in schema.",
+                )
+            })?
+            .field_type
+            .as_ref()
+        else {
+            return Err(Error::new(
+                crate::ErrorKind::Unexpected,
+                "Composed type is not supported for min max aggregation.",
+            ));
+        };
+
+        macro_rules! update_stat {
+            ($self:ident, $stat:ident, $convert_func:expr) => {
+                if $stat.min_is_exact() {
+                    let val = $convert_func($stat.min().clone())?;
+                    match $self.lower_bounds.entry(col_id) {
```

Comment:
Is there no way to do something similar this instead, which would be a bit neater? I've not tried it, so I may be wrong 😅 

```rust
$self.lower_bounds.entry(col_id)
    .and_modify(|e| *e.min(&val))
    .or_insert(val);
```